contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -198,26 +214,65 @@ contract LidoOracle is ILidoOracle, AragonApp {
      * @param _beaconValidators Number of validators visible on this epoch
      */
     function reportBeacon(uint256 _epochId, uint128 _beaconBalance, uint128 _beaconValidators) external {
-        (uint256 startEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
-        require(_epochId >= startEpoch, "EPOCH_IS_TOO_OLD");
+        BeaconSpec memory beaconSpec = _getBeaconSpec();
+        (uint256 currentEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
+        require(_epochId >= currentEpoch, "EPOCH_IS_TOO_OLD");
         require(_epochId <= endEpoch, "EPOCH_HAS_NOT_YET_BEGUN");
+        require(_epochId.mod(beaconSpec.epochsPerFrame) == 0, "MUST_BE_FIRST_EPOCH_FRAME");
 
-        address member = msg.sender;
-        uint256 index = _getMemberId(member);
-        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        // If reported epoch advanced, clear last unsuccessfull voting
+        if (_epochId > currentEpoch) _clearVotingAndAdvanceTo(_epochId);
 
-        // check & set contribution flag
-        uint256 bitMask = gatheredEpochData[_epochId].reportsBitMask;
+        // Make sure the oracle is from members list and has not yet voted
+        uint256 index = _getMemberId(msg.sender);
+        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        uint256 bitMask = REPORTS_BITMASK_POSITION.getStorageUint256();
         require(!bitMask.getBit(index), "ALREADY_SUBMITTED");
+        REPORTS_BITMASK_POSITION.setStorageUint256(bitMask.setBit(index, true));
```

Comment:
Is it possible that we'll not have enough bits in uint256?

Reply:
no, max number of oracles is 
```
 uint256 public constant MAX_MEMBERS = 256;
```